mkosi/machine.py

```diff
@@ -122,9 +142,70 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
         if self._serial:
             self._serial.kill(signal.SIGTERM)
             self.exit_code = self._serial.wait()
             self._serial = None
         self.stack.__exit__(*args, **kwargs)
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self._kill(*args, **kwargs)
+
+    def reboot(self) -> None:
+        self._kill(None, None, None)
+        self._boot()
+
+
+class MkosiMachineTest(unittest.TestCase):
+    args: Sequence[str]
+    machine: Machine
+    exc_info: Tuple[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]] \
+              = (None, None, None)
+
+    def _callTestMethod(self, method: str) -> None:
+        try:
+            super()._callTestMethod(method) # type: ignore
+        except Exception as e:
+            self.exc_info = exc_info()
+            raise e
+
+    @classmethod
+    def _parse_boolean(cls, s: Optional[str]) -> bool:
+        "Parse 1/true/yes/y/t/on as true and 0/false/no/n/f/off/None as false"
+        if not s:
+            return False
+
+        s_l = s.lower()
+        if s_l in {"1", "true", "yes", "y", "t", "on"}:
+            return True
+
+        if s_l in {"0", "false", "no", "n", "f", "off"}:
+            return False
+
+        raise ValueError(f"Invalid literal for bool(): {s!r}")
```

Comment:
I just grossly misunderstood what you said, my bad.

Reply:
The thing is parse_boolean requires a value, while `os.getenv()` returns an `Optional[sequence]`.
If we pass an optional arg, then we can't use the `s_l = s.lower()`, since we cant lower() None.
So this would probably break this function as to how it works right now.

Reply:
I just grossly misunderstood what you said. The downside here is that this kinda changes how `parse_boolean`() works by making it return False if the string is empty.
This would raise a valueError originally.

We could check if the strings are different than None before calling the function, but we would still need to change the argument type and use the `if s`

Reply:
We could do this:
`if not os.getenv("MKOSI_TESTS_NO_NSPAWN"):
            no_nspawn = False
        else:
            no_nspawn = parse_boolean(os.getenv("MKOSI_TESTS_NO_NSPAWN"))

        if not os.getenv("MKOSI_TESTS_NO_NSPAWN"):
            no_qemu = False
        else:
            no_qemu = parse_boolean(os.getenv("MKOSI_TESTS_NO_QEMU"))`

And then parse_boolean would look like this:
`def parse_boolean(s: Optional[str]) -> bool:
    "Parse 1/true/yes/y/t/on as true and 0/false/no/n/f/off/None as false"
    if s:
        s_l = s.lower()

    if s_l in {"1", "true", "yes", "y", "t", "on"}:
        return True

    if s_l in {"0", "false", "no", "n", "f", "off"}:
        return False

    raise ValueError(f"Invalid literal for bool(): {s!r}")
`

So its behavior remains the same.

Reply:
You can configure the default value that `getenv` should return instead of `None`

So what you could do is `no_nspawn = parse_boolean(os.getenv("MKOSI_TESTS_NO_NSPAWN", "0"))`